plugins/base/src/main/java/edu/wpi/first/shuffleboard/plugin/base/widget/GraphWidget.java

```diff
@@ -282,111 +247,110 @@ private void updateFromArraySource(DataSource<double[]> source) {
-    // Note this only affects the chart; the actual data is not changed
-    Data<Number, Number> squarifier = null;
-    double prev = queue.get(queue.size() - 1).getYValue().doubleValue();
-    if (prev != newData) {
-      squarifier = new Data<>(elapsed - 1, prev);
+      data.add(elapsed, nextValue);
+    });
+
+    if (!chart.getDatasets().contains(data)
+        && Optional.ofNullable(visibleSeries.get(data)).map(Property::getValue).orElse(true)) {
+      chart.getDatasets().add(data);
     }
-    return squarifier;
   }
 
-  private Series<Number, Number> getNumberSeries(DataSource<? extends Number> source) {
+  private DoubleDataSet getNumberSeries(DataSource<? extends Number> source) {
     if (!numberSeriesMap.containsKey(source)) {
-      Series<Number, Number> series = new Series<>();
-      series.setName(source.getName());
+      DoubleDataSet series = new DoubleDataSet(source.getName());
       numberSeriesMap.put(source, series);
-      realData.put(series, new SimpleData());
       visibleSeries.computeIfAbsent(series, createVisibleProperty);
-      series.nodeProperty().addListener((__, old, node) -> {
-        if (node instanceof Parent) {
-          Parent parent = (Parent) node;
-          parent.getChildrenUnmodifiable().forEach(child -> {
-            child.setCache(true);
-            child.setCacheHint(CacheHint.SPEED);
-          });
-          parent.setCache(true);
-          parent.setCacheHint(CacheHint.SPEED);
-        }
-      });
     }
     return numberSeriesMap.get(source);
   }
 
-  private List<Series<Number, Number>> getArraySeries(DataSource<double[]> source) {
-    List<Series<Number, Number>> series = arraySeriesMap.computeIfAbsent(source, __ -> new ArrayList<>());
+  private List<DoubleDataSet> getArraySeries(DataSource<double[]> source) {
+    List<DoubleDataSet> series = arraySeriesMap.computeIfAbsent(source, __ -> new ArrayList<>());
     final double[] data = source.getData();
     if (data.length < series.size()) {
       while (series.size() != data.length) {
-        Series<Number, Number> removed = series.remove(series.size() - 1);
-        realData.remove(removed);
+        DoubleDataSet removed = series.remove(series.size() - 1);
         visibleSeries.remove(removed);
       }
     } else if (data.length > series.size()) {
       for (int i = series.size(); i < data.length; i++) {
-        Series<Number, Number> newSeries = new Series<>();
-        newSeries.setName(source.getName() + "[" + i + "]"); // eg "array[0]", "array[1]", etc
+        DoubleDataSet newSeries = new DoubleDataSet(source.getName() + "[" + i + "]");
         series.add(newSeries);
-        realData.put(newSeries, new SimpleData());
         visibleSeries.computeIfAbsent(newSeries, createVisibleProperty);
       }
     }
     return series;
   }
 
-  private void updateBounds(long elapsed) {
-    xAxis.setUpperBound(elapsed);
-    removeInvisibleData();
-  }
-
   private void update() {
     FxUtils.runOnFxThread(() -> {
```

Comment:
This is way too much code to live in a single lambda. Split it up into discrete functions